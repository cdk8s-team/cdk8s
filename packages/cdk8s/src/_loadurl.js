// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

const { http, https } = require('follow-redirects');
const { parse } = require('url');
const fs = require('fs');

const url = process.argv[2];
if (!url) {
  console.error(`Usage: ${process.argv[1]} URL`);
  process.exit(1);
}

const purl = parse(url);

if (!purl.protocol) {
  if (!purl.pathname) {
    throw new Error(`unable to parse pathname from file url: ${url}`);
  }

  process.stdout.write(fs.readFileSync(purl.pathname, 'utf-8'));
  process.exit(0);
}

const client = getHttpClient();

const get = client.get(url, response => {
  if (response.statusCode !== 200) {
    throw new Error(`${response.statusCode} response from http get: ${response.statusMessage}`);
  }

  response.on('data', chunk => process.stdout.write(chunk));
});

get.once('error', err => {
  throw new Error(`http error: ${err.message}`);
});

function getHttpClient() {
  switch (purl.protocol) {
    case 'http:': return http;
    case 'https:': return https;
    default:
      throw new Error(`unsupported protocol "${purl.protocol}" in url: ${url}`);
  }
}
