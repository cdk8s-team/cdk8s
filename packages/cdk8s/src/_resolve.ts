import { Lazy } from './lazy';

export function resolve(value: any): any {

  if (value == null) {
    return value;
  }

  // cdk8s token
  if (value instanceof Lazy) {
    const resolved = value.produce();
    return resolve(resolved);
  }

  // implicit token as generated by the cdk8s-cli (e.g IntOrString)
  if (typeof(value.resolve) === 'function') {
    const resolved = value.resolve();
    return resolve(resolved);
  }

  if (typeof(value) !== 'object') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(x => resolve(x));
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    result[k] = resolve(v);
  }

  return result;

}
