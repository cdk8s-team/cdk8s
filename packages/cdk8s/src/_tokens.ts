import { Tokenization, DefaultTokenResolver, StringConcat, Construct } from 'constructs';

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

export function resolve<T>(scope: Construct, obj: T): T {
  return Tokenization.resolve(obj, {
    scope,
    resolver: TOKEN_RESOLVER,
  });
}