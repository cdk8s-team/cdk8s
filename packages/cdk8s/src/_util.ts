export interface SanitizeOptions {
  /**
   * Do not include empty objects (no keys).
   */
  readonly filterEmptyObjects?: boolean;

  /**
   * Do not include arrays with no items.
   */
  readonly filterEmptyArrays?: boolean;
}

export function sanitizeValue(obj: any, options: SanitizeOptions = { }): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {

    if (options.filterEmptyArrays && obj.length === 0) {
      return undefined;
    }

    return obj.map(x => sanitizeValue(x, options));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = sanitizeValue(value, options);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  if (options.filterEmptyObjects && Object.keys(newObj).length === 0) {
    return undefined;
  }

  return newObj;
}
