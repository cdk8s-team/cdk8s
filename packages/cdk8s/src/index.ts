export * from './api-object';
export * from './chart';
export * from './testing';
export * from './app';
export * from './include';
export * from './yaml';
export * from './metadata';