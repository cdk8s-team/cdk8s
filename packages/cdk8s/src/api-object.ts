import { Construct, IConstruct, Node } from 'constructs';
import { Chart } from './chart';
import { sanitizeValue } from './_util';
import { resolve } from './_tokens';
import * as stringify from 'json-stable-stringify';
import { ApiObjectMetadata, ApiObjectMetadataDefinition } from './metadata';

/**
 * Options for defining API objects.
 */
export interface ApiObjectOptions {
  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {
  /**
   * The name of the API object.
   *
   * If a name is specified in `metadata.name` this will be the name returned.
   * Otherwise, a name will be generated by calling
   * `Chart.of(this).generatedObjectName(this)`, which by default uses the
   * construct path to generate a DNS-compatible name for the resource.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  /**
   * The chart in which this object is defined.
   */
  public readonly chart: Chart;

  /**
   * Metadata associated with this API object.
   */
  public readonly metadata: ApiObjectMetadataDefinition;

  /**
   * Defines an API object.
   *
   * @param scope the construct scope
   * @param ns namespace
   * @param options options
   */
  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.chart = Chart.of(this);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;

    this.name = options.metadata?.name ?? this.chart.generateObjectName(this);

    this.metadata = new ApiObjectMetadataDefinition({
      name: this.name,
      namespace: options.metadata?.namespace ?? this.chart.namespace,

      // override user defined values
      ...options.metadata,
    });
  }

  /**
   * Create a dependency between this ApiObject and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const data = {
      ...this.options,
      metadata: this.metadata.toJson(),
    };

    // convert to "pure data" so, for example, when we convert to yaml these
    // references are not converted to anchors.
    return JSON.parse(stringify(sanitizeValue(resolve(this, data))));
  }
}
