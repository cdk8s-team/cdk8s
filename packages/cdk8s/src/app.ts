import * as fs from 'fs';
import * as path from 'path';
import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { Chart } from './chart';
import { DependencyGraph } from './dependency';
import { Names } from './names';
import { Yaml } from './yaml';

export interface AppProps {
  /**
   * The directory to output Kubernetes manifests.
   *
   * @default - CDK8S_OUTDIR if defined, otherwise "dist"
   */
  readonly outdir?: string;
}

/**
 * Represents a cdk8s application.
 */
export class App extends Construct {
  /**
   * Synthesize a single chart.
   *
   * Each element returned in the resulting array represents a different ApiObject
   * in the scope of the chart.
   *
   * Note that the returned array order is important. It is determined by the various dependencies between
   * the constructs in the chart, where the first element is the one without dependencies, and so on...
   *
   * @returns An array of JSON objects.
   * @param chart the chart to synthesize.
   * @internal
   */
  public static _synthChart(chart: Chart): any[] {

    const app: App = App.of(chart);

    // we must prepare the entire app before synthesizing the chart
    // because the dependency inference happens on the app level.
    resolveDependencies(app);

    // validate the app since we want to call onValidate of the relevant constructs.
    // note this will also call onValidate on constructs from possibly different charts,
    // but thats ok too since we no longer treat constructs as a self-contained synthesis unit.
    validate(app);

    return chartToKube(chart);
  }

  private static of(c: IConstruct): App {

    const scope = Node.of(c).scope;

    if (!scope) {
      // the app is the only construct without a scope.
      return c as App;
    }

    return App.of(scope);
  }

  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /**
   * Defines an app
   * @param props configuration options
   */
  constructor(props: AppProps = { }) {
    super(undefined as any, '');
    this.outdir = props.outdir ?? process.env.CDK8S_OUTDIR ?? 'dist';
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {

    fs.mkdirSync(this.outdir, { recursive: true });

    // this is kind of sucky, eventually I would like the DependencyGraph
    // to be able to answer this question.
    const hasDependantCharts = resolveDependencies(this);

    // Since we plan on removing the distributed synth mechanism, we no longer call `Node.synthesize`, but rather simply implement
    // the necessary operations. We do however want to preserve the distributed validation.
    validate(this);

    const simpleManifestNamer = (chart: Chart) => `${Names.toDnsLabel(Node.of(chart).path)}.k8s.yaml`;
    const manifestNamer = hasDependantCharts ? (chart: Chart) => `${index.toString().padStart(4, '0')}-${simpleManifestNamer(chart)}` : simpleManifestNamer;

    const charts: IConstruct[] = new DependencyGraph(Node.of(this)).topology().filter(x => x instanceof Chart);

    let index = 0;
    for (const node of charts) {
      const chart: Chart = Chart.of(node);
      Yaml.save(path.join(this.outdir, manifestNamer(chart)), chartToKube(chart));
      index++;
    }

  }
}

function validate(app: App) {

  // Note this is a copy-paste of https://github.com/aws/constructs/blob/master/lib/construct.ts#L438.
  const errors = Node.of(app).validate();
  if (errors.length > 0) {
    const errorList = errors.map(e => `[${Node.of(e.source).path}] ${e.message}`).join('\n  ');
    throw new Error(`Validation failed with the following errors:\n  ${errorList}`);
  }

}

function resolveDependencies(app: App) {

  let hasDependantCharts = false;

  for (const dep of Node.of(app).dependencies) {

    // create explicit api object dependencies from implicit construct dependencies
    const targetApiObjects = Node.of(dep.target).findAll().filter(c => c instanceof ApiObject);
    const sourceApiObjects = Node.of(dep.source).findAll().filter(c => c instanceof ApiObject);

    for (const target of targetApiObjects) {
      for (const source of sourceApiObjects) {
        Node.of(source).addDependency(target);
      }
    }

    // create an explicit chart dependency from implicit construct dependencies
    const sourceChart = Chart.of(dep.source);
    const targetChart = Chart.of(dep.target);

    if (sourceChart !== targetChart) {
      Node.of(sourceChart).addDependency(targetChart);
      hasDependantCharts = true;
    }

  }

  return hasDependantCharts;

}

function chartToKube(chart: Chart) {
  return new DependencyGraph(Node.of(chart)).topology()
    .filter(x => x instanceof ApiObject)
    .map(x => (x as ApiObject).toJson());
}
