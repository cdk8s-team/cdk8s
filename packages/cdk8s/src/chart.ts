import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { App } from './app';
import { Names } from './names';

export interface ChartProps {
  /**
   * The default namespace for all objects defined in this chart (directly or
   * indirectly). This namespace will only apply to objects that don't have a
   * `namespace` explicitly defined for them.
   *
   * @default - no namespace is synthesized (usually this implies "default")
   */
  readonly namespace?: string;

  /**
   * Labels to apply to all resources in this chart.
   *
   * @default - no common labels
   */
  readonly labels?: { [name: string]: string };
}

export class Chart extends Construct {

  /**
   * Finds the chart in which a node is defined.
   * @param c a construct node
   */
  public static of(c: IConstruct): Chart {
    if (c instanceof Chart) {
      return c;
    }

    const parent = Node.of(c).scope as Construct;
    if (!parent) {
      throw new Error('cannot find a parent chart (directly or indirectly)');
    }

    return Chart.of(parent);
  }

  /**
   * The default namespace for all objects in this chart.
   */
  public readonly namespace?: string;

  /**
   * Chart-level labels.
   */
  private readonly _labels?: { [name: string]: string };

  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id);
    this.namespace = props.namespace;
    this._labels = props.labels ?? {};
  }

  /**
   * Labels applied to all resources in this chart.
   *
   * This is an immutable copy.
   */
  public get labels(): { [name: string]: string } {
    return { ...this._labels };
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * Different resource types may have different constraints on names
   * (`metadata.name`). The previous version of the name generator was
   * compatible with DNS_SUBDOMAIN but not with DNS_LABEL.
   *
   * For example, `Deployment` names must comply with DNS_SUBDOMAIN while
   * `Service` names must comply with DNS_LABEL.
   *
   * Since there is no formal specification for this, the default name
   * generation scheme for kubernetes objects in cdk8s was changed to DNS_LABEL,
   * since itâ€™s the common denominator for all kubernetes resources
   * (supposedly).
   *
   * You can override this method if you wish to customize object names at the
   * chart level.
   *
   * @param apiObject The API object to generate a name for.
   */
  public generateObjectName(apiObject: ApiObject) {
    return Names.toDnsLabel(apiObject);
  }

  /**
   * Create a dependency between this Chart and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders this chart to a set of Kubernetes JSON resources.
   * @returns array of resource manifests
   */
  public toJson(): any[] {
    return App._synthChart(this);
  }
}

