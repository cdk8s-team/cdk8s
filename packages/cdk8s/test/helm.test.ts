import * as child_process from 'child_process';
import * as path from 'path';
import { Helm, Testing } from '../src';

const SAMPLE_CHART_PATH = path.join(__dirname, 'fixtures', 'helm-sample');

beforeEach(() => jest.restoreAllMocks());

test('basic usage', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  const helm = new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
  });

  // THEN
  expect(helm.releaseName).toEqual('test-sample-00742e24');
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('fails if the helm executable is not found', () => {
  // GIVEN
  const chart = Testing.chart();
  const executable = `foo-${Math.random() * 999999}`;

  // THEN
  expect(() => new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
    helmExecutable: executable,
  })).toThrow(`unable to execute '${executable}' to render Helm chart`);
});

test('values can be specified when defining the chart', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
    values: {
      replicaCount: 889,
      ingress: {
        enabled: false,
      },
      nodeSelector: {
        selectMe: 'boomboom',
      },
    },
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('releaseName can be used to specify the release name', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  const helm = new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
    releaseName: 'your-release',
  });

  // THEN - all names start with "your-release-"
  expect(helm.releaseName).toEqual('your-release');
  const names: string[] = Testing.synth(chart).map(obj => obj.metadata.name);
  for (const n of names) {
    expect(n.startsWith('your-release-')).toBeTruthy();
  }
});

test('it is possible to interact with api objects in the chart', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  const helm = new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
  });

  const service = helm.apiObjects.find(o => o.kind === 'ServiceAccount' && o.name === 'test-sample-00742e24-helm-sample');
  service?.metadata.addAnnotation('my.annotation', 'hey-there');

  // THEN
  expect(helm.apiObjects.map(o => `${o.kind}/${o.name}`).sort()).toEqual([
    'Deployment/test-sample-00742e24-helm-sample',
    'Service/test-sample-00742e24-helm-sample',
    'ServiceAccount/test-sample-00742e24-helm-sample',
  ]);

  expect(service?.toJson().metadata.annotations).toEqual({
    'my.annotation': 'hey-there',
  });
});

test('helmFlags can be used to specify additional helm options', () => {
  // GIVEN
  const spawnMock = jest.spyOn(child_process, 'spawnSync').mockReturnValue({
    status: 0,
    stderr: Buffer.from(''),
    stdout: Buffer.from(''),
    pid: 123,
    output: ['stdout', 'stderr'],
    signal: null,
  });

  const chart = Testing.chart();

  // WHEN
  new Helm(chart, 'sample', {
    chart: SAMPLE_CHART_PATH,
    helmFlags: [
      '--description', 'my custom description',
      '--no-hooks',
    ],
  });

  // THEN
  const expectedArgs: string[] = [
    'template',
    '--description', 'my custom description',
    '--no-hooks',
    'test-sample-00742e24',
    SAMPLE_CHART_PATH,
  ];

  expect(spawnMock).toHaveBeenCalledTimes(1);
  expect(spawnMock).toHaveBeenCalledWith('helm', expectedArgs);
});

test('propagates helm failures', () => {
  // GIVEN
  const chart = Testing.chart();

  // THEN
  expect(() => new Helm(chart, 'my-chart', {
    chart: SAMPLE_CHART_PATH,
    helmFlags: ['--invalid-argument-not-found-boom-boom'],
  })).toThrow(/unknown flag/);
});