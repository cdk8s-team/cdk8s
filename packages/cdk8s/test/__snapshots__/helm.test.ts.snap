// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic usage 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 80,
          "protocol": "TCP",
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/name": "helm-sample",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "test-sample-c8e2763d",
          "app.kubernetes.io/name": "helm-sample",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "test-sample-c8e2763d",
            "app.kubernetes.io/name": "helm-sample",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "nginx:1.16.0",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "name": "helm-sample",
              "ports": Array [
                Object {
                  "containerPort": 80,
                  "name": "http",
                  "protocol": "TCP",
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "resources": Object {},
              "securityContext": Object {},
            },
          ],
          "securityContext": Object {},
          "serviceAccountName": "test-sample-c8e2763d-helm-sample",
        },
      },
    },
  },
]
`;

exports[`values can be specified when defining the chart 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 80,
          "protocol": "TCP",
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/name": "helm-sample",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-sample-c8e2763d",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "helm-sample",
        "app.kubernetes.io/version": "1.16.0",
        "helm.sh/chart": "helm-sample-0.1.0",
      },
      "name": "test-sample-c8e2763d-helm-sample",
    },
    "spec": Object {
      "replicas": 889,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "test-sample-c8e2763d",
          "app.kubernetes.io/name": "helm-sample",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "test-sample-c8e2763d",
            "app.kubernetes.io/name": "helm-sample",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "nginx:1.16.0",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "name": "helm-sample",
              "ports": Array [
                Object {
                  "containerPort": 80,
                  "name": "http",
                  "protocol": "TCP",
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "resources": Object {},
              "securityContext": Object {},
            },
          ],
          "nodeSelector": Object {
            "selectMe": "boomboom",
          },
          "securityContext": Object {},
          "serviceAccountName": "test-sample-c8e2763d-helm-sample",
        },
      },
    },
  },
]
`;
